core/common/src/grpc_interceptor.rs

```diff
@@ -0,0 +1,238 @@
+                service_name = qualified_service_name_parts[1].to_string();
+            }
+        }
+        (service_name, method_name)
+    }
+}
+
+impl<S> Service<http::request::Request<tonic::transport::Body>> for GrpcInterceptorService<S>
+where
+    S: Service<
+            http::request::Request<tonic::transport::Body>,
+            Response = http::response::Response<tonic::body::BoxBody>,
+            Error = Box<dyn std::error::Error + Sync + Send>,
+        > + Send,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    /// Implementation of tower's Service trait's poll_ready method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    /// Implementation of tower's Service trait's call method.
+    /// See https://docs.rs/tower/latest/tower/trait.Service.html
+    fn call(
+        &mut self,
+        mut request: http::request::Request<tonic::transport::Body>,
+    ) -> Self::Future {
+        let interceptor = (self.interceptor_factory)();
+
+        let (service_name, method_name) = Self::retrieve_grpc_names_from_uri(request.uri());
+        let is_applicable = interceptor.is_applicable(&service_name, &method_name);
+
+        if is_applicable && interceptor.must_handle_request() {
+            let (parts, body) = request.into_parts();
+            let mut body_bytes: Bytes =
+                match futures::executor::block_on(hyper::body::to_bytes(body)) {
+                    Ok(bytes) => bytes,
+                    Err(err) => {
+                        return Box::pin(async move {
+                            Err(Box::new(err) as Box<dyn std::error::Error + Sync + Send>)
+                        })
+                    }
+                };
+            let protobuf_message_bytes: Bytes = body_bytes.split_off(GRPC_HEADER_LENGTH);
+            let grpc_header_bytes = body_bytes;
+            let new_protobuf_message_bytes: Bytes = match interceptor.handle_request(
+                &service_name,
+                &method_name,
+                protobuf_message_bytes,
+            ) {
+                Ok(bytes) => bytes,
+                Err(err) => return Box::pin(async move { Err(err) }),
+            };
+            let new_body_chunks: Vec<Result<_, std::io::Error>> =
+                vec![Ok(grpc_header_bytes), Ok(new_protobuf_message_bytes)];
+            let stream = futures_util::stream::iter(new_body_chunks);
+            let new_body = tonic::transport::Body::wrap_stream(stream);
+            request = http::request::Request::from_parts(parts, new_body);
+        }
+
+        let fut = self.service.call(request);
+
+        Box::pin(async move {
+            match fut.await {
+                Ok(response) => {
```

Comment:
Consider using if let and unwrap here to reduce the big match statement.

`
let result = fut.await;

if let Err(err) = result { 
  return Err(err);
}

let ok_response = result.unwrap()
....
`

Reply:
May be able to get away with something like `let response = fut.await?;` or `let response = fut.await.map_err(|err| => Err(err))?;

Reply:
I think this would work too: let `let ok_response = fut.await?;`

Reply:
I second @devkelley's suggestion and would write this as follows:

```rust
Box::pin(async move {
    let mut response = fut.await?;
    if is_applicable && interceptor.must_handle_response() {
        // lines 201-226 go here
        response = http::response::Response::from_parts(parts, new_box_body);
    }

    Ok(response)
})
```

Reply:
I'm fine with the way that it is.